src/Features/CSharp/Portable/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,438 @@
+                editor.ReplaceNode(containingPattern, rewrittenPattern);
+                return editor.GetChangedRoot();
+            };
+        }
+
+        private static PatternSyntax RewriteContainingPattern(
+            PatternSyntax containingPattern,
+            PatternSyntax generatedPattern,
+            ImmutableArray<IdentifierNameSyntax> namesOpt)
+        {
+            PatternSyntax result;
+            if (namesOpt.IsDefault)
+            {
+                // If there's no name, this is a variable designation match.
+                result = (containingPattern, generatedPattern) switch
+                {
+                    // We know we have a var-pattern, declaration-pattern or a recursive-pattern on the left as the containing node of the variable designation.
+                    // Depending on the generated pattern off of the expression on the right, we can give a better result by morphing it into the existing match.
+                    // Otherwise, we fallback to an `and` pattern.
+                    (VarPatternSyntax var, RecursivePatternSyntax recurisve) => recurisve.WithDesignation(var.Designation),
+                    (DeclarationPatternSyntax decl, RecursivePatternSyntax recurisve) => recurisve.WithType(decl.Type).WithDesignation(decl.Designation),
+                    (RecursivePatternSyntax recurisve, DeclarationPatternSyntax decl) => recurisve.WithType(decl.Type),
+                    _ => BinaryPattern(AndPattern, containingPattern, generatedPattern),
+                };
+            }
+            else
+            {
+                // Otherwise, generate a subpattern per each name.
+                var subpattern = CreateSubpattern(namesOpt, generatedPattern);
+                result = containingPattern switch
+                {
+                    VarPatternSyntax p => RecursivePattern(subpattern, p.Designation),
+                    DeclarationPatternSyntax p => RecursivePattern(subpattern, p.Designation, p.Type),
+                    RecursivePatternSyntax p => p.AddPropertyPatternClauseSubpatterns(subpattern),
+                    var p => throw ExceptionUtilities.UnexpectedValue(p)
+                };
+            }
+
+            // We must have preserved the existing variable designation.
+            Debug.Assert(containingPattern switch
+            {
+                VarPatternSyntax p => p.Designation,
+                DeclarationPatternSyntax p => p.Designation,
+                RecursivePatternSyntax p => p.Designation,
+                var p => throw ExceptionUtilities.UnexpectedValue(p)
+            } is var d && result.DescendantNodes().Any(node => AreEquivalent(node, d)));
+
+            return result.WithAdditionalAnnotations(Formatter.Annotation);
+        }
+
+        private static PatternSyntax CreatePattern(ExpressionSyntax receiver, ExpressionOrPatternSyntax target, bool flipped)
```

Comment:
what does 'flipped' generally mean here?

Reply:
"Should we flip a relational operator in a relational pattern?" e.g. `0 < e.x`  should be converted to `{ x: > 0 }`